pkg/reconciler/v1alpha1/clusteringress/resources/secret.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/knative/serving/pkg/apis/networking"
+	"github.com/knative/serving/pkg/apis/networking/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+)
+
+// GetSecrets gets the all of the secrets referenced by the given ClusterIngress, and
+// returns a map whose key is the a secret namespace/name key and value is pointer of the secret.
+func GetSecrets(ci *v1alpha1.ClusterIngress, secretLister corev1listers.SecretLister) (map[string]*corev1.Secret, error) {
+	secrets := map[string]*corev1.Secret{}
+	for _, tls := range ci.Spec.TLS {
+		ref := fmt.Sprintf("%s/%s", tls.SecretNamespace, tls.SecretName)
+		if _, ok := secrets[fmt.Sprintf("%s/%s", tls.SecretNamespace, tls.SecretName)]; ok {
+			continue
+		}
+		secret, err := secretLister.Secrets(tls.SecretNamespace).Get(tls.SecretName)
+		if err != nil {
+			return nil, err
+		}
+		secrets[ref] = secret
+	}
+	return secrets, nil
+}
+
+// MakeSecrets makes copies of the origin Secrets under the namespace of Istio gateway service.
+func MakeSecrets(ctx context.Context, originSecrets map[string]*corev1.Secret) []*corev1.Secret {
+	gatewaySvcNamespaces := getAllGatewaySvcNamespaces(ctx)
+	secrets := []*corev1.Secret{}
+	for _, originSecret := range originSecrets {
+		for _, ns := range gatewaySvcNamespaces {
+			if ns == originSecret.Namespace {
+				// no need to copy secret when the target namespace is the same
+				// as the origin namespace
+				continue
+			}
+			secrets = append(secrets, makeSecret(originSecret, ns))
+		}
+	}
+	return secrets
+}
+
+func makeSecret(originSecret *corev1.Secret, targetNamespace string) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      targetSecret(originSecret),
+			Namespace: targetNamespace,
+			Labels: map[string]string{
+				networking.OriginSecretNameLabelKey:      originSecret.Name,
```

Comment:
do we care about the other labels?

Reply:
we don't care about other labels.
There are some `cert-manager` specific labels in the origin secrets. They are used by cert-manager controller to find out the generated secrets for the cert-manager `Certificate` objects. We don't need them in the target secrets. Actually adding them to the target secrets could potentially make cert-manager controller get the wrong secrets for `Certificate` objects.